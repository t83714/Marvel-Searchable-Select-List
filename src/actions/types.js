export const APP_INIT = "APP_INIT";
export const LOADING_ICON_STATUS_UPDATE = "LOADING_ICON_STATUS_UPDATE";
export const OPTION_DATA_UPDATE = "OPTION_DATA_UPDATE";
export const USER_INPUT_CHANGE = "USER_INPUT_CHANGE";
export const OPTION_SELECTED = "OPTION_SELECTED";
