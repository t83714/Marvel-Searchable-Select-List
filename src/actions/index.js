import {
    APP_INIT,
    USER_INPUT_CHANGE,
    OPTION_DATA_UPDATE,
    LOADING_ICON_STATUS_UPDATE,
    OPTION_SELECTED,
} from "./types";

export function appInit() {
    return {
        type: APP_INIT,
        payload: null,
    };
}

export function userInputChange(inputValue) {
    return {
        type: USER_INPUT_CHANGE,
        payload: inputValue,
    };
}

export function optionDataUpdate(options) {
    return {
        type: OPTION_DATA_UPDATE,
        payload: options,
    };
}

export function loadingIconStatusUpdate(status) {
    return {
        type: LOADING_ICON_STATUS_UPDATE,
        payload: status,
    };
}

export function optionSelected(id) {
    return {
        type: OPTION_SELECTED,
        payload: id,
    };
}
