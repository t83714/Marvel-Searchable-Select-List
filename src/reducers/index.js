import combineReducers from "../common/combineReducers";
import optionsReducer, { init as optionDataInit } from "./options";
import {
    LOADING_ICON_STATUS_UPDATE,
    USER_INPUT_CHANGE,
} from "../actions/types";

export const init = function () {
    return {
        options: optionDataInit(),
        isLoadingIconShown: false,
        userInput: "",
    };
};

const AppInitState = init();

// eslint-disable-next-line no-unused-vars
const globalReducer = function (state = AppInitState, action) {
    switch (action.type) {
    case LOADING_ICON_STATUS_UPDATE:
        return {
            ...state,
            isLoadingIconShown: action.payload,
        };
    case USER_INPUT_CHANGE:
        return {
            ...state,
            userInput: action.payload,
        };
    default: return state;
    }
};


const combinedReducer = combineReducers({
    options: optionsReducer,
});

const AppReducer = function (state = AppInitState, action) {
    let newState = globalReducer(state, action);
    newState = combinedReducer(newState, action);
    return newState;
};

export default AppReducer;
