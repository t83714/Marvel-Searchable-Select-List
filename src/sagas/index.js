import { takeEvery, takeLatest } from "redux-saga/effects";
import { USER_INPUT_CHANGE } from "../actions/types";
import loadOptions from "./loadOptions";

// eslint-disable-next-line no-unused-vars
function* AppSaga(store) {
    yield takeLatest(USER_INPUT_CHANGE, loadOptions);
}

export default AppSaga;
