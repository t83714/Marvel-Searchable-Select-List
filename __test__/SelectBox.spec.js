import React from "react";

import 'core-js/es6/map';
import 'core-js/es6/set';

import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import { shallow, mount } from "enzyme";
import renderer from "react-test-renderer";
import ConnectedSearchBox, { SearchBox } from "../src/views/SearchBox";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

import { optionDataUpdate, loadingIconStatusUpdate, optionSelected } from "../src/actions";
import { createStore } from "redux";
import appReducers, { init as appInit } from "../src/reducers";

const raf = global.requestAnimationFrame = function (callback) {
    setTimeout(callback, 0);
};

configure({ adapter: new Adapter() });

// Snapshot for SelectBox Component
describe(">>>SearchBox --- Snapshot", () => {
    it("+++capturing Snapshot of SearchBox", () => {
        const state = appInit();
        state.options = [
            {
                id: "123",
                name: "Iron Man 1",
                description: "sdfsa wsdfonaire industrialisdsfinstead created an advanced suit of armor to save his life and escape captivity. Now with a new outlook on life, Tony uses his money and intelligence to make the world a safer, better place as Iron Man.",
                thumbnail: {
                    path: "http://i.annihil.us/u/prod/marvel/i/mg/9/c0/527bb7b37ff55",
                    extension: "jpg",
                },
            },
            {
                id: "89798",
                name: "Iron Man 2",
                description: "Wousdfred and forceddf by his enemies, billionaire industrialist Tony Stark instead created an advanced suit of armor to save his life and escape captivity. Now with a new outlook on life, Tony uses his money and intelligence to make the world a safer, better place as Iron Man.",
                thumbnail: {
                    path: "http://i.annihil.us/u/prod/marvel/i/mg/9/c0/527bb7b37ff55",
                    extension: "jpg",
                },
            },
            {
                id: "233",
                name: "Iron Man 3",
                description: "Wousdfdred and forced tosdfis enemies, billionaire industrialist Tony Stark instead created an advanced suit of armor to save his life and escape captivity. Now with a new outlook on life, Tony uses his money and intelligence to make the world a safer, better place as Iron Man.",
                thumbnail: {
                    path: "http://i.annihil.us/u/prod/marvel/i/mg/9/c0/527bb7b37ff55",
                    extension: "jpg",
                },
            },
        ];
        const renderedValue = renderer.create(<SearchBox {...state} />).toJSON();
        expect(renderedValue).toMatchSnapshot();
    });
});
//* ******************************************************************************************************
describe(">>>SearchBox--- Shallow Render REACT COMPONENTS", () => {
    let wrapper;
    const state = appInit();

    beforeEach(() => {
        wrapper = shallow(<SearchBox {...state} />);
    });

    it("+++ render the DUMB component", () => {
        expect(wrapper.length).toEqual(1);
    });

    it("+++ contains input", () => {
        expect(wrapper.find("input").length).toBe(1);
    });
});


//* ******************************************************************************************************
describe(">>>SearchBox --- REACT-REDUX (Shallow + passing the {store} directly)", () => {
    const initialState = appInit();
    const mockStore = configureStore();
    let store,
        container;

    beforeEach(() => {
        store = mockStore(initialState);
        container = shallow(<ConnectedSearchBox store={store} />);
    });

    it("+++ render the connected(SMART) component", () => {
        expect(container.length).toEqual(1);
    });

    it("+++ check Prop matches with initialState", () => {
        expect(container.prop("isLoadingIconShown")).toEqual(initialState.isLoadingIconShown);
    });
});

//* ******************************************************************************************************
describe(">>>SearchBox --- REACT-REDUX (Mount + wrapping in <Provider>)", () => {
    const initialState = appInit();
    const mockStore = configureStore();
    let store,
        wrapper;

    beforeEach(() => {
        store = mockStore(initialState);
        wrapper = mount(<Provider store={store}><ConnectedSearchBox /></Provider>);
    });

    it("+++ contains input", () => {
        expect(wrapper.find("input").length).toBe(1);
    });

    it("+++ render the connected(SMART) component", () => {
        expect(wrapper.find(ConnectedSearchBox).length).toEqual(1);
    });

    it("+++ check Prop matches with initialState", () => {
        expect(wrapper.find(SearchBox).prop("isLoadingIconShown")).toEqual(initialState.isLoadingIconShown);
    });

    it("+++ check action on dispatching ", () => {
        let action;
        store.dispatch(loadingIconStatusUpdate(true));
        store.dispatch(optionSelected("32332"));
        action = store.getActions();
        expect(action[0].type).toBe("LOADING_ICON_STATUS_UPDATE");
        expect(action[1].type).toBe("OPTION_SELECTED");
    });
});

//* ******************************************************************************************************
describe(">>>SearchBox --- REACT-REDUX (actual Store + reducers) more of Integration Testing", () => {
    const initialState = appInit();
    let store,
        wrapper;

    beforeEach(() => {
        store = createStore(appReducers);
        wrapper = mount(<Provider store={store}><ConnectedSearchBox /></Provider>);
    });

    it("+++ check Prop matches with initialState", () => {
        store.dispatch(loadingIconStatusUpdate(true));
        expect(wrapper.find(SearchBox).length).toBe(1);
    });
});
//* ******************************************************************************************************
