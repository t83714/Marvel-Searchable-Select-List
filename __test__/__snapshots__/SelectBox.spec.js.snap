// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>SearchBox --- Snapshot +++capturing Snapshot of SearchBox 1`] = `
<form
  className="select-box form-inline my-2 my-lg-0"
>
  <div />
  <input
    aria-label="Search"
    className="form-control mr-sm-2"
    onChange={[Function]}
    placeholder="Search"
    type="text"
  />
</form>
`;
